pennylane/ops/op_math/decompositions/controlled_decompositions.py

```diff
@@ -349,6 +369,200 @@ def controlled_two_qubit_unitary_rule(U, wires, control_values, work_wires, **__
+
+
+def _two_workers_resource(num_control_wires, work_wire_type, **__):
+    if work_wire_type == "clean":
+        n_ccx = 2 * num_control_wires - 3
+        return {ops.Toffoli: n_ccx, ops.X: n_ccx - 3 if num_control_wires < 6 else n_ccx - 5}
+    # Otherwise, we assume the work wires are dirty
+    n_ccx = 4 * num_control_wires - 8
+    return {ops.Toffoli: n_ccx, ops.X: n_ccx - 4 if num_control_wires < 6 else n_ccx - 8}
+
+
+@register_condition(_two_workers_condition)
+@register_resources(_two_workers_resource)
+def _decompose_mcx_with_two_workers(wires, work_wires, work_wire_type, **__):
+    r"""
+    Synthesise a multi-controlled X gate with :math:`k` controls using :math:`2` ancillary qubits.
+    It produces a circuit with :math:`2k-3` Toffoli gates and depth :math:`O(\log(k))` if using
+    clean ancillae, and :math:`4k-8` Toffoli gates and depth :math:`O(\log(k))` if using dirty
+    ancillae as described in Sec. 5 of [1].
+
+    References:
+        1. Khattar and Gidney, Rise of conditionally clean ancillae for optimizing quantum circuits
+        `arXiv:2407.17966 <https://arxiv.org/abs/2407.17966>`__
+
+    """
+
+    # First use the work wire to prepare the first two control wires as conditionally clean.
+    ops.Toffoli([wires[0], wires[1], work_wires[0]])
+    middle_ctrl_indices = _build_log_n_depth_ccx_ladder(wires[:-1])
+
+    # Apply the MCX in the middle
+    if len(middle_ctrl_indices) == 1:
+        ops.Toffoli([work_wires[0], wires[middle_ctrl_indices[0]], wires[-1]])
+    else:
+        middle_wires = [wires[i] for i in middle_ctrl_indices]
+        _decompose_mcx_with_one_worker(work_wires[:1] + middle_wires + wires[-1:], work_wires[1:])
+
+    # Uncompute the first ladder
+    ops.adjoint(_build_log_n_depth_ccx_ladder, lazy=False)(wires[:-1])
+    ops.Toffoli([wires[0], wires[1], work_wires[0]])
+
+    if work_wire_type == "dirty":
+        # Perform toggle-detection of the work wire is dirty
+        middle_ctrl_indices = _build_log_n_depth_ccx_ladder(wires[:-1])
+        if len(middle_ctrl_indices) == 1:
+            ops.Toffoli([work_wires[0], wires[middle_ctrl_indices[0]], wires[-1]])
+        else:
+            middle_wires = [wires[i] for i in middle_ctrl_indices]
+            _decompose_mcx_with_one_worker(
+                work_wires[:1] + middle_wires + wires[-1:], work_wires[1:]
+            )
+        ops.adjoint(_build_log_n_depth_ccx_ladder, lazy=False)(wires[:-1])
+
+
+decompose_mcx_with_two_workers = flip_zero_control(_decompose_mcx_with_two_workers)
+
+
+def _decompose_mcx_one_worker_condition(num_control_wires, num_work_wires, **__):
+    return num_control_wires > 2 and num_work_wires == 1
+
+
+def _decompose_mcx_one_worker_resource(num_control_wires, work_wire_type, **__):
+    if work_wire_type == "clean":
+        n_ccx = 2 * num_control_wires - 3
+        return {ops.Toffoli: n_ccx, ops.X: n_ccx - 3}
+    # Otherwise, we assume the work wire is dirty
+    n_ccx = 4 * num_control_wires - 8
```

Comment:
Wouldn't there be less ops if work wires are dirty since we don't need to uncompute? Or am I misunderstanding the meaning of the work wire type?

Reply:
No, `dirty` work wires means that they are not zeroed at the start, but they still need to be uncomputed.